doc/proposals/fast_ipam_by_node.md

```diff
@@ -0,0 +1,271 @@
+
+We will create a new CRD `NodeSlicePool` which will be used to manage the slices of the network ranges that nodes
+are assigned to. A controller where reconcile these NodeSlicePools based on cluster nodes and network-attachment-definitions.
+
+Whereabouts binary will be able to tell this feature is enabled and when creating `IPPools` it will check the `NodeSlicePool` to get the range of the current node.
+It will set this on existing IPPools object and use a lease per IPPool. There will be an `IPPool` and `Lease` per network per node.
+Where a network is defined by network name i.e. you can have multiple `network-attachment-definitions` with a shared network name and this will result in
+a shared `NodeSlicePool`, `IPPool` and `Lease` per node for these `network-attachment-definitions`.
+
+i.e. we have nad1 and nad2 both with network name `test-network`. When a node, `trusted-otter` joins the cluster this will result in
+`NodeSlicePool`, `IPPool` and `Lease` objects named `test-network-trusted-otter`. If these are seperate network you would just not set the network name
+or set the network-name differently per `network-attachment-defintion`.
+
+### Changes in IPAM Config
+
+This will lead to change in IPAM config something as follows:
+
+<table>
+<tr>
+<th>Old IPAM Config</th>
+<th>Changes</th>
+<th>New IPAM Config</th>
+</tr>
+<tr>
+<td>
+  
+```json
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28"
+      }
+}
+```
+  
+</td>
+<td>
+
+```diff
+{
+      "cniVersion": "0.3.0",
+      "name": "whereaboutsexample",
+      "type": "macvlan",
+      "master": "eth0",
+      "mode": "bridge",
+      "ipam": {
+        "type": "whereabouts",
+        "range": "192.168.2.225/28"
++       "fast_ipam": true,
++       "node_slice_size": "/22"
+      }
+}
+```
+
+</td>
+<td>
+
+```json
+{
+  "cniVersion": "0.3.0",
+  "name": "whereaboutsexample",
+  "type": "macvlan",
+  "master": "eth0",
+  "mode": "bridge",
+  "ipam": {
+    "type": "whereabouts",
+    "range": "192.168.2.225/28",
+    "fast_ipam": true,
+    "node_slice_size": "/22"
+  }
+}
+```
+
+</td>
+</tr>
+</table>
+
+_Note: But at the same time we also need to support the old field (i.e. field duplication) in order to make the design backward compatible._
+
+### Changes in Modules
+
+#### whereabouts/pkg/types/types.go
+
+```diff
+ type IPAMConfig struct {
+      Name                string
+      Type                string               `json:"type"`
+      Routes              []*cnitypes.Route    `json:"routes"`
+      Datastore           string               `json:"datastore"`
+      IPRanges            []RangeConfiguration `json:"ipRanges"`
++     FastIpam            bool                 `json:"fast_ipam"`
```

Comment:
is the boolean really required ? 

Can't we simply just infer if fast IPAM is used if the node slice size is provisioned ?

Reply:
Yeah sure we could do that

Reply:
updated per this feedback, ty I'll add this to the prototype today